libraries/standard/mqtt/src/mqtt_lightweight.c

```diff
@@ -252,6 +270,167 @@ static int32_t recvExact( MQTTTransportRecvFunc_t recvFunc,
 
 /*-----------------------------------------------------------*/
 
+static bool publishPacketSize( const MQTTPublishInfo_t * pPublishInfo,
+                               size_t * pRemainingLength,
+                               size_t * pPacketSize )
+{
+    bool status = true;
+    size_t packetSize = 0, payloadLimit = 0;
+
+    assert( pPublishInfo != NULL );
+    assert( pRemainingLength != NULL );
+    assert( pPacketSize != NULL );
+
+    /* The variable header of a PUBLISH packet always contains the topic name.
+     * The first 2 bytes of UTF-8 string contains length of the string.
+     */
+    packetSize += pPublishInfo->topicNameLength + sizeof( uint16_t );
+
+    /* The variable header of a QoS 1 or 2 PUBLISH packet contains a 2-byte
+     * packet identifier. */
+    if( pPublishInfo->qos > MQTTQoS0 )
+    {
+        packetSize += sizeof( uint16_t );
+    }
+
+    /* Calculate the maximum allowed size of the payload for the given parameters.
+     * This calculation excludes the "Remaining length" encoding, whose size is not
+     * yet known. */
+    payloadLimit = MQTT_MAX_REMAINING_LENGTH - packetSize - 1U;
+
+    /* Ensure that the given payload fits within the calculated limit. */
+    if( pPublishInfo->payloadLength > payloadLimit )
+    {
+        IotLogErrorWithArgs( "Publish packet length exceeds the MQTT maximum length." );
```

Comment:
This log message can be improved to include expected and actual maximum length. Also, for messages with no args `IotLogError` should be used - please check this at other places too.